internal/controllers/app_controller.go

```diff
@@ -270,12 +290,245 @@ func (r *AppReconciler) reconcile(ctx context.Context, app *ketchv1.App, logger
+				recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "%s", stringifyEvent(msg))
+			}
+		case <-healthcheckTimeout:
+			err = createDeployTimeoutError(ctx, cli, app, time.Since(now), namespace, "healthcheck")
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "error waiting for healthcheck: %s", err.Error())
+			return
+		case <-timeout:
+			err = createDeployTimeoutError(ctx, cli, app, time.Since(now), namespace, "full rollout")
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "deployment timeout: %s", err.Error())
+			return
+		case <-ctx.Done():
+			return
+		}
+	}()
+
+	for {
+		for i := range dep.Status.Conditions {
+			c := dep.Status.Conditions[i]
+			if c.Type == DeploymentProgressing && c.Reason == deadlineExeceededProgressCond {
+				recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "deployment %q exceeded its progress deadline", dep.Name)
+				return errors.Errorf("deployment %q exceeded its progress deadline", dep.Name)
+			}
+		}
+		if oldUpdatedReplicas != dep.Status.UpdatedReplicas {
+			recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "%d of %d new units created", dep.Status.UpdatedReplicas, specReplicas)
+		}
+
+		if healthcheckTimeout == nil && dep.Status.UpdatedReplicas == specReplicas {
+			err := checkPodStatus(r.Group, r.Client, app.Name, app.Spec.Deployments[len(app.Spec.Deployments)-1].Version)
+			if err == nil {
+				healthcheckTimeout = time.After(maxWaitTimeDuration)
+				recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "waiting healthcheck on %d created units", specReplicas)
+			}
+		}
+
+		readyUnits := dep.Status.UpdatedReplicas - dep.Status.UnavailableReplicas
+		if oldReadyUnits != readyUnits && readyUnits >= 0 {
+			recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "%d of %d new units ready", readyUnits, specReplicas)
+		}
+
+		pendingTermination := dep.Status.Replicas - dep.Status.UpdatedReplicas
+		if oldPendingTermination != pendingTermination && pendingTermination > 0 {
+			recorder.Eventf(app, v1.EventTypeNormal, appReconcileUpdate, "%d old units pending termination", pendingTermination)
+		}
+
+		oldUpdatedReplicas = dep.Status.UpdatedReplicas
+		oldReadyUnits = readyUnits
+		oldPendingTermination = pendingTermination
+		if readyUnits == specReplicas &&
+			dep.Status.Replicas == specReplicas {
+			break
+		}
+
+		dep, err = cli.AppsV1().Deployments(namespace).Get(context.TODO(), dep.Name, metav1.GetOptions{})
+		if err != nil {
+			recorder.Eventf(app, v1.EventTypeWarning, appReconcileError, "error getting deployments: %s", err.Error())
+			return err
+		}
+	}
+
+	outcome := ketchv1.AppReconcileOutcome{AppName: app.Name, DeploymentCount: int(dep.Status.ReadyReplicas)}
+	recorder.Event(app, v1.EventTypeNormal, appReconcileComplete, outcome.String())
+	return nil
+}
+
+// stringifyEvent accepts an event and returns relevant details as a string
+func stringifyEvent(watchEvent watch.Event) string {
```

Comment:
maybe using annotations, similar to CanaryEvents is better solution, what do you think?

Reply:
That sounds good. I removed this and added an `AppDeploymentEvent` type that includes `annotations` and `String()`, similar to the Canary work. Hoping it's close to what you are expecting.